### PR TITLE
Bob can decline a swap request

### LINKED ISSUE #381: Implement "decline" action on started state for Bob
- [x] Implement http API call to trigger "decline" action
- [x] Bob sends message to Alice, similar to Accept

Test assertions:
- [x] Bob API returns the swap state as "declined"
- [x] Alice API returns the swap state as "declined"

Child of #405.

### PR DESCRIPTION
**Pull request by luckysori**
Resolves #381.

### CODE REVIEW

File: application/comit_node/src/comit_client/bam.rs [hunk 1]
```diff
@@ -88,11 +88,27 @@ impl Client for BamClient {
...
+                        match serde_json::from_value(response.get_body().clone()) {
                             Ok(response) => Ok(Ok(response)),
                             Err(_e) => Err(SwapResponseError::InvalidResponse),
                         }
                     }
+                    Status::SE(20) => {
+                        info!("{} declined swap request: {:?}", socket_addr, response);
+                        Ok(Err({
+                            let reason = response
+                                .get_header("REASON")
```
Comment (thomaseizinger) [inline application/comit_node/src/comit_client/bam.rs]: I would appreciate it if we can discuss changes to the COMIT protocol first before they are implemented!
ATM, it is completely undocumented. We should flesh out first, how we want things to be before they are implemented.

Regarding this particular change: I don't see a usecase for custom messages provided by the user. Those are only an attack vector for spamming or possible exploits through buffer overflows in other implementations.

If we want to provide reasons for rejections, I would suggest to discuss and specify a limited set of possible reasons (BAD_RATE for example). Everything else is useless in regards to interoperability and automation which is what we are striving for.

PS: I'll try not to work from now on :)

Comment (luckysori) [reply application/comit_node/src/comit_client/bam.rs]: This was just my interpretation of this comment: `// TODO: Implement SwapReject::Decline(reason)`.

But I completely agree with you. I have decided instead to keep the field as an enum, which is something that I documented in one of the commit messages as a future improvement (and what you suggest). For now the enum only has one variant `Unspecified`.

Thank you for the comment, but please do stop working :laughing:

Comment (D4nte) [reply application/comit_node/src/comit_client/bam.rs]: "I would appreciate it if we can discuss changes to the COMIT protocol first before they are implemented!" -> you do understand that we may do changes to the protocol while you are away? 😇

File: api_tests/dry/rfc003/test.js [hunk 2]
```diff
@@ -62,7 +62,7 @@ describe("RFC003 HTTP API", () => {
             .post('/swaps/rfc003')
             .send({
                 "alpha_ledger": {
-                    "name": "Thomas' wallet",
```
Comment (D4nte) [inline api_tests/dry/rfc003/test.js]: rebase and format :)

File: application/comit_node/src/comit_client/bam.rs (hunk 1 above)
Comment (D4nte) [reply application/comit_node/src/comit_client/bam.rs]: Having written the `TODO` comment and had some verbal discussions with the team I realize that I did not express myself clearly enough.

I was proposing to actually have a range of Error code `Status::SE` that could express defined/common reason of rejections such as:
- `RateDeclined`
- `UnsupportedLedger`
- `InternalError`
- `MissingLiquidity` (this is an idea, not feasible at this stage).

Do we agree that `Status::SE` is to be defined by the application level (ie, COMIT being an application level for BAM).

File: application/comit_node/src/comit_client/mod.rs [hunk 3]
```diff
@@ -31,10 +31,18 @@ pub trait ClientFactory<C: Client>: Send + Sync + RefUnwindSafe + Debug {
...
+    Declined {
+        reason: SwapDeclineReason,
+    },
     /// The counterparty rejected the request
     Rejected,
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub enum SwapDeclineReason {
+    Unspecified,
```
Comment (D4nte) [inline application/comit_node/src/comit_client/mod.rs]: Shall we implement few reasons the user can pass in the `POST` request?
Also, is there some existing `internal error` cases that we could clean up?
When do we use `Rejected`? is that internal errors only?
I think under `Rejected` we could also have an enum. Something like:
```
pub enum SwapReject {
    Declined {
        pub enum Reason {
            Liquidity,
            Rate,
            Unspecified,
            AlphaTimeout,
            BetaTimeout,
        }
    },
    Rejected {
        pub enum Reason {
            UnsupportedLedger,
            InternalError,
        }
    }
}
```

Comment (D4nte) [reply application/comit_node/src/comit_client/mod.rs]: Maybe worth doing a quick meeting/brainstorm about it.

File: application/comit_node/src/comit_client/bam.rs (hunk 1 above)
Comment (LLFourn) [reply application/comit_node/src/comit_client/bam.rs]: @D4nte The rfc says that `SE20`  and above are reserved for the application.

File: application/comit_node/src/comit_client/mod.rs (hunk 3 above)
Comment (LLFourn) [reply application/comit_node/src/comit_client/mod.rs]: @D4nte I think that idea has potential.

File: application/comit_node/src/comit_client/bam.rs (hunk 1 above)
Comment (D4nte) [reply application/comit_node/src/comit_client/bam.rs]: ok cool, so I guess we can use those for COMIT :)

File: application/comit_node/src/comit_client/mod.rs (hunk 3 above)
Comment (D4nte) [reply application/comit_node/src/comit_client/mod.rs]: #583 opened to track

**Pull request merged**